File: app/src/main/java/com/letianpai/robot/control/callback/GestureCallback.java
package com.letianpai.robot.control.callback;


import com.renhejia.robot.gesturefactory.parser.GestureData;

import java.util.ArrayList;

/**
 * @author liujunbin
 */
public class GestureCallback {


    private ArrayList<GestureResponseListener> mGestureResponseListener = new ArrayList<>();
    private ArrayList<GestureCompleteListener> mGestureCompleteListener = new ArrayList<>();
    private ArrayList<GestureCompleteListener> mOneShotGestureCompleteListener = new ArrayList<>();

    private static class GestureCallbackHolder {
        private static final GestureCallback instance = new GestureCallback();
    }

    private GestureCallback() {

    }

    public static GestureCallback getInstance() {
        return GestureCallbackHolder.instance;
    }

    public interface GestureResponseListener {
        void onGestureReceived(String gesture);

        void onGestureReceived(String gesture, int gestureId);

        void onGesturesReceived(ArrayList<GestureData> list, int taskId);

        void onGesturesReceived(GestureData gestureData);
    }

    public void setGestureListener(GestureResponseListener listener) {
        this.mGestureResponseListener.add(listener);
    }

    public void setGestureCompleteListener(GestureCompleteListener listener) {
        this.mGestureCompleteListener.add(listener);
    }

    public void setOneShotGestureCompleteListener(GestureCompleteListener mOneShotGestureCompleteListener) {
        this.mOneShotGestureCompleteListener.add(mOneShotGestureCompleteListener);
    }

    public void removeOneShotGestureCompleteListener(GestureCompleteListener mGestureCompleteListener) {
        if (this.mOneShotGestureCompleteListener.size() > 0) {
            mOneShotGestureCompleteListener.remove(mGestureCompleteListener);
        }
    }


    public void setGesture(String gesture) {
        for (int i = 0; i < mGestureCompleteListener.size(); i++) {
            if (mGestureResponseListener.get(i) != null) {
                mGestureResponseListener.get(i).onGestureReceived(gesture);
            }
        }
    }

    public void setGesture(String gesture, int geTaskId) {
        for (int i = 0; i < mGestureResponseListener.size(); i++) {
            if (mGestureResponseListener.get(i) != null) {
                mGestureResponseListener.get(i).onGestureReceived(gesture);
            }
        }
    }

    public void setGestures(ArrayList<GestureData> list, int taskId) {
        for (int i = 0; i < mGestureResponseListener.size(); i++) {
            if (mGestureResponseListener.get(i) != null) {
                mGestureResponseListener.get(i).onGesturesReceived(list, taskId);
            }
        }
    }

    public void setGesture(GestureData gestureData) {
        for (int i = 0; i < mGestureResponseListener.size(); i++) {
            if (mGestureResponseListener.get(i) != null) {
                mGestureResponseListener.get(i).onGesturesReceived(gestureData);
            }
        }
    }

    public interface GestureCompleteListener {
        void onGestureCompleted(String gesture, int geTaskId);
    }

    public void setGesturesComplete(String gesture, int geTaskId) {
        for (int i = 0; i < mGestureResponseListener.size(); i++) {
            if (mGestureCompleteListener.get(i) != null) {
                mGestureCompleteListener.get(i).onGestureCompleted(gesture, geTaskId);
            }
        }
        for (int i = 0; i < mOneShotGestureCompleteListener.size(); i++) {
            if (mOneShotGestureCompleteListener.get(i) != null) {
                mOneShotGestureCompleteListener.get(i).onGestureCompleted(gesture, geTaskId);
            }
        }
    }
}

File: app/src/main/java/com/letianpai/robot/control/callback/GestureCallback.kt
package com.letianpai.robot.control.callback

import com.renhejia.robot.gesturefactory.parser.GestureData


/**
 * @author liujunbin
 */
class GestureCallback private constructor() {
    private val mGestureResponseListener = ArrayList<GestureResponseListener?>()
    private val mGestureCompleteListener = ArrayList<GestureCompleteListener?>()
    private val mOneShotGestureCompleteListener = ArrayList<GestureCompleteListener?>()

    private object GestureCallbackHolder {
        val instance: GestureCallback = GestureCallback()
    }

    interface GestureResponseListener {
        fun onGestureReceived(gesture: String?)

        fun onGestureReceived(gesture: String, gestureId: Int)

        fun onGesturesReceived(list: ArrayList<GestureData>, taskId: Int)

        fun onGesturesReceived(gestureData: GestureData?)
    }

    fun setGestureListener(listener: GestureResponseListener?) {
        mGestureResponseListener.add(listener)
    }

    fun setGestureCompleteListener(listener: GestureCompleteListener?) {
        mGestureCompleteListener.add(listener)
    }

    fun setOneShotGestureCompleteListener(mOneShotGestureCompleteListener: GestureCompleteListener?) {
        this.mOneShotGestureCompleteListener.add(mOneShotGestureCompleteListener)
    }

    fun removeOneShotGestureCompleteListener(mGestureCompleteListener: GestureCompleteListener?) {
        if (mOneShotGestureCompleteListener.size > 0) {
            mOneShotGestureCompleteListener.remove(mGestureCompleteListener)
        }
    }


    fun setGesture(gesture: String?) {
        for (i in mGestureCompleteListener.indices) {
            if (mGestureResponseListener[i] != null) {
                mGestureResponseListener[i]!!.onGestureReceived(gesture)
            }
        }
    }

    fun setGesture(gesture: String?, geTaskId: Int) {
        for (i in mGestureResponseListener.indices) {
            if (mGestureResponseListener[i] != null) {
                mGestureResponseListener[i]!!.onGestureReceived(gesture)
            }
        }
    }

    fun setGestures(list: ArrayList<GestureData>, taskId: Int) {
        for (i in mGestureResponseListener.indices) {
            if (mGestureResponseListener[i] != null) {
                mGestureResponseListener[i]!!.onGesturesReceived(list, taskId)
            }
        }
    }

    fun setGesture(gestureData: GestureData?) {
        for (i in mGestureResponseListener.indices) {
            if (mGestureResponseListener[i] != null) {
                mGestureResponseListener[i]!!.onGesturesReceived(gestureData)
            }
        }
    }

    fun interface GestureCompleteListener {
        fun onGestureCompleted(gesture: String, geTaskId: Int)
    }

    fun setGesturesComplete(gesture: String, geTaskId: Int) {
        for (i in mGestureResponseListener.indices) {
            if (mGestureCompleteListener[i] != null) {
                mGestureCompleteListener[i]!!.onGestureCompleted(gesture, geTaskId)
            }
        }
        for (i in mOneShotGestureCompleteListener.indices) {
            if (mOneShotGestureCompleteListener[i] != null) {
                mOneShotGestureCompleteListener[i]!!.onGestureCompleted(gesture, geTaskId)
            }
        }
    }

    companion object {
        val instance: GestureCallback
            get() = GestureCallbackHolder.instance
    }
}
